// @modxclub/old__/view/office/sections/companies/index.js
import React, {Component} from 'react';

import PropTypes from 'prop-types';

import Grid from 'material-ui/Grid';
import Typography from 'material-ui/Typography';

import List, {
  ListItem,
  ListItemIcon,
  ListItemText,
  ListItemSecondaryAction,
  ListSubheader,
} from 'material-ui/List';

import {Link} from 'react-router-dom';

import OfficeSectionLayout from '../layout';

import CompanyCard from './company';

import WarningMessage from '../../../../fields/warning';

export default class OfficeCompaniesSection extends OfficeSectionLayout{


	static contextTypes = {
		companiesStore: PropTypes.object.isRequired,
	};
 


 	getMenu(){

    let {
      user: {
        user,
      }
    } = this.props;

    let {
      username,
    } = user || {};

    return <List> 

      <Link
        to={`/office/`}
        href={`/office/`}
      >
        <ListItem
          button
        > 
          В начало
        </ListItem>
      </Link>
    </List>
  } 

  renderDefaultBlock(){

    let {
      companiesStore,
    } = this.context;

    let content;

    // console.log('companiesStore.getState()', companiesStore.getState());

    if(!companiesStore.getState().size){
      content = <Typography>
        У вас нет ни одной команды. <Link
          to={`/office/company/create/`}
          href={`/office/company/create/`}
        >
          Создайте свою первую команду.
        </Link>
      </Typography>;
    }



    return content;
  }



  getPageHeader(pagetitle){

    // return super.getPageHeader("Управление wefew");
  }

  setPagetitle(title){
 

    // return super.setPagetitle("Управление wefewf");
  }

 	renderContent(){

    let {
      router: {
        params: {
          // section,
          action,
          id,
        },
      },
    } = this.props;

    let {
    	companiesStore,
    } = this.context;

    let content;

    switch(action){

      case 'create':

        content = <CompanyCard 
          companiesStore={companiesStore}
        />

        break;

    	case 'edit':

        if(!parseInt(id)){

          return <WarningMessage>
            Не был получен ID
          </WarningMessage>;
        }

        let item = companiesStore.getState().find(n => n.company_id == id);

        // console.log("companiesStore", companiesStore.getState());
        // console.log("item", id, item);

        if(!item){

          return <WarningMessage>
            Не была получена команда
          </WarningMessage>;
        }

  		  return <CompanyCard 
          companiesStore={companiesStore}
    			item={item}
    		/>

    		break;

    	default: content = this.renderDefaultBlock();
    }

    return content;
 	}
}
