// @modxclub/old__/menu/index.jsx
import React, {Component} from 'react';
import { createStyleSheet } from 'jss-theme-reactor';
import customPropTypes from 'material-ui/utils/customPropTypes';
import List, {
  ListItem,
  ListItemText,
  ListItemIcon,
  ListItemSecondaryAction
} from 'material-ui/List';
import Divider from 'material-ui/Divider';
import IconButton from 'material-ui/IconButton';
import InboxIcon from 'material-ui-icons/Inbox';
import Home from 'material-ui-icons/Home';
import Chat from 'material-ui-icons/Chat';
import LibraryBooks from 'material-ui-icons/LibraryBooks';
import Extension from 'material-ui-icons/Extension';
import ImportantDevices from 'material-ui-icons/ImportantDevices';
import Contacts from 'material-ui-icons/MailOutline';
import InfoOutline from 'material-ui-icons/InfoOutline';
import TrendingUp from 'material-ui-icons/TrendingUp';
import KeyboardArrowLeft from 'material-ui-icons/KeyboardArrowLeft';
import KeyboardArrowRight from 'material-ui-icons/KeyboardArrowRight';
import PeopleIcon from 'material-ui-icons/People';
import CompaniesIcon from 'material-ui-icons/Business';
import BlogsIcon from 'material-ui-icons/FolderOpen';
import CommentsIcon from 'material-ui-icons/Forum';
import {Link} from 'react-router-dom';

const iconStyle = {
  width: 24,
  height: 24,
  marginRight: 0,
  flex: 'none',
}

const styleSheet = createStyleSheet('MainMenu', (theme) => ({
  root: {
    width: '100%',
    background: theme.palette.background.paper,
    position: 'relative',

    '& a': {

      textDecoration: 'none',
    }
  },
  // rootDocked: {
  // },
}));

var classes;

export default class MainMenu extends Component{

  SwitchPage(url){

    this.props.documentActions.LoadDocument(url);
  }

  componentWillMount(){

    classes = this.context.styleManager.render(styleSheet);
  }

  render(){

    return (
      <div className={[this.props.className, classes.root, classes.rootDocked].join(" ")}>
        <List>

          <Link to='/' href='/'>
            <ListItem
              button
            >
              <ListItemIcon>
                <Home  
                  style={iconStyle}
                />
              </ListItemIcon>
              <ListItemText primary="Главная" />

              {/*
                <ListItemSecondaryAction
                  className="menu-toggler"
                >
                  <IconButton
                    onTouchTap={event => {
                      event.stopPropagation();
                      event.preventDefault();
                      this.props.onToggleClick(event);
                    }}
                  >
                    <KeyboardArrowLeft />
                  </IconButton>
                </ListItemSecondaryAction>
              */}
            </ListItem>
          </Link>

          {/*
            <ListItem
              button
              onTouchTap={() => {
                this.SwitchPage('/chat/');
              }}
            >
              <ListItemIcon>
                <Chat />
              </ListItemIcon>
              <ListItemText primary="Чат" />
            </ListItem>
          */}

          <Link to='/blogs/' href='/blogs/'>
            <ListItem
              button
            >
              <ListItemIcon>
                <BlogsIcon  
                  style={iconStyle}
                />
              </ListItemIcon>
              <ListItemText primary="Блоги" />
            </ListItem>
          </Link>

          <Link to='/topics/' href='/topics/'>
            <ListItem
              button
            >
              <ListItemIcon>
                <LibraryBooks  
                  style={iconStyle}
                />
              </ListItemIcon>
              <ListItemText primary="Топики" />
            </ListItem>
          </Link>

          <Link to='/comments/' href='/comments/'>
            <ListItem
              button
            >
              <ListItemIcon>
                <CommentsIcon  
                  style={iconStyle}
                />
              </ListItemIcon>
              <ListItemText primary="Комментарии" />
            </ListItem>
          </Link>

          <Link to='/downloads/' href='/downloads/'>
            <ListItem
              button
            >
              <ListItemIcon>
                <Extension 
                  style={iconStyle}
                />
              </ListItemIcon>
              <ListItemText primary="Дополнения для MODX" />
            </ListItem>
          </Link>

          <Link to='/katalog-sajtov/' href='/katalog-sajtov/'>
            <ListItem
              button
            >
              <ListItemIcon>
                <ImportantDevices  
                  style={iconStyle}
                />
              </ListItemIcon>
              <ListItemText primary="Каталог сайтов" />
            </ListItem>
          </Link>

          <Link to='/katalog-web-studij/' href='/katalog-web-studij/'>
            <ListItem
              button
            >
              <ListItemIcon>
                <CompaniesIcon  
                  style={iconStyle}
                />
              </ListItemIcon>
              <ListItemText primary="Каталог веб-студий" />
            </ListItem>
          </Link>

          <Link to='/people/' href='/people/'>
            <ListItem
              button
            >
              <ListItemIcon>
                <PeopleIcon  
                  style={iconStyle}
                />
              </ListItemIcon>
              <ListItemText primary="Пользователи" />
            </ListItem>
          </Link>

          <Link to='/contacts.html' href='/contacts.html'>
            <ListItem
              button
            >
                <ListItemIcon>
                  <Contacts  
                  style={iconStyle}
                />
                </ListItemIcon>
                <ListItemText primary="Контакты" />
            </ListItem>
          </Link>

        </List>

        <Divider
          className="menu--menu-bottom-border"
        />
      </div>
    );

        // <Divider />

    // return <Menu
    //   open={true}
    // >
    //   <MenuItem onTouchTap={this.SwitchPage.bind(this, '/')}>Главная</MenuItem>
    //   <MenuItem onTouchTap={this.SwitchPage.bind(this, '/chat/')}>Чат</MenuItem>
    //   <MenuItem onTouchTap={this.SwitchPage.bind(this, '/topics/')}>Топики</MenuItem>
    //   <MenuItem onTouchTap={this.SwitchPage.bind(this, '/downloads/')}>Дополнения для MODX</MenuItem>
    //   <MenuItem onTouchTap={this.SwitchPage.bind(this, '/katalog-sajtov/')}>Каталог сайтов</MenuItem>
    //   <MenuItem onTouchTap={this.SwitchPage.bind(this, '/contacts.html')}>Контакты</MenuItem>
    // </Menu>
  }
}

MainMenu.contextTypes = {
  styleManager: customPropTypes.muiRequired,
};
