// @modxclub/query/index.js

import gql from "graphql-tag";

import {
  UserNoNestingFragment,
  NotificationTypeNoNestingFragment,
  EthTransactionNoNestingFragment,
} from "../../../schema/generated/api.fragments";


export const updateUserProcessor = gql`


  mutation updateUserProcessor(
    $data: UserUpdateInput!
    $where: UserWhereUniqueInput
  ){
    response: updateUserProcessor(
      where: $where
      data: $data
    ){
      success
      message
      errors{
        key
        message
      }
      data{
        ...UserNoNesting
        NotificationTypes{
          ...NotificationTypeNoNesting
        }
      }
    }
  }


  ${UserNoNestingFragment}
  ${NotificationTypeNoNestingFragment}

`;


export const createEthTransactionProcessor = gql`
  mutation createEthTransactionProcessor(
    $data: EthTransactionCreateInput!
  ){
    response: createEthTransactionProcessor(
      data: $data
    ){
      success
      message
      errors{
        key
        message
      }
      data{
        ...EthTransactionNoNesting
      }
    }
  }

  ${EthTransactionNoNestingFragment}
`;
