// @modxclub/old__/appbar/index.jsx

import React from 'react';
import ReactDOM   from 'react-dom';

import customPropTypes from 'material-ui/utils/customPropTypes';

import { createStyleSheet } from 'jss-theme-reactor';

import AppBar from 'material-ui/AppBar';
import Toolbar from 'material-ui/Toolbar';
import Typography from 'material-ui/Typography';
import Button from 'material-ui/Button';
import IconButton from 'material-ui/IconButton';
import Menu, {MenuItem} from 'material-ui-icons/Menu';

import Dialog from 'material-ui/Dialog';
import Badge from 'material-ui/Badge';

import Layout from 'material-ui/Grid';


import Notifications from '../notifications';
import ProfileIcon from '../profile/';
import Balance from '../balance/';
import Search from '../search/';

import MenuIcon from 'material-ui-icons/Menu';
 

import Proxy from '../proxy';

const propTypes = {
};

const defaultProps = {
  showMenu: true,
};


const styleSheet = createStyleSheet('ButtonAppBar', () => ({
  root: {
    position: 'sticky',
    width: '100%',
    top: 0,
  },
  appBar: {
    position: 'relative',
    flexDirection: 'row',
    padding: '5px 0',
    transition: 'padding 0.5s',
    boxShadow: 'none',
    '&.minified':{
      padding: '0',
    }
  },
  flex: {
    display: "flex",
    alignItems: "center",
  },
  first_block: {
    flex: 1,
    padding: "0 15px",
  },
}));

var classes;

// console.log('classes', classes);


class MainAppBar extends React.Component{

  constructor(props){
    super(props);

    this.state = {
      appBarClass: "minified",
    }
  }


  handleLogin(){
    // this.props.userActions.loginClicked();
    // setTimeout(this.props.userActions.loginComplete(),3000);
    this.props.userActions.loginComplete();
  }

  // SimpleAppBar(props, context) {
  //   const classes = context.styleManager.render(styleSheet);
  //
  //   console.log('classes', classes);
  //
  //   return (
  //     <div className={classes.root}>
  //       <AppBar className={classes.appBar}>
  //         <Toolbar>
  //           <Typography type="title" colorInherit>Title</Typography>
  //         </Toolbar>
  //       </AppBar>
  //     </div>
  //   );
  // }

  // onScroll(){
  //   if(typeof window != "undefined"){
  //     var last_known_scroll_position = window.scrollY;
  //     // console.log('last_known_scroll_position', last_known_scroll_position, this);

  //     var appBarClass = null;

  //     if(last_known_scroll_position > 200){
  //       appBarClass = "minified";
  //       if(this.state.appBarClass !== appBarClass){
  //         this.setState({
  //           appBarClass,
  //         });
  //       }
  //     }
  //     else{
  //       if(
  //         this.state.appBarClass == "minified"
  //         && last_known_scroll_position < 100
  //       ){
  //         appBarClass = "";
  //         if(this.state.appBarClass !== appBarClass){
  //           this.setState({
  //             appBarClass,
  //           });
  //         }
  //       }
  //     }
  //     // else if(last_known_scroll_position > 60){
  //     //   root_class = "root_transited";
  //     // }

  //   }
  // }


  componentWillMount(){
    classes = this.context.styleManager.render(styleSheet);
  }

  // componentDidMount(){
  //   if(typeof window != "undefined"){
  //     window.addEventListener('scroll', () => {
  //       this.onScroll();
  //     });
  //   }
  // }

  render() {

    var rootClasses = ["appbar", classes.root];

    var appBarClass = [classes.appBar];

    if(this.state.appBarClass != ""){
      appBarClass.push(this.state.appBarClass);
    }

    return (
      <div className={rootClasses.join(" ")}>
        <AppBar
          className={appBarClass.join(" ")}
          // style={{ backgroundImage: 'linear-gradient(to left,#C82C60,#8033A5)', boxShadow: 'none' }}
        >
          <div className={[classes.flex, classes.first_block].join(" ")}>

            {this.props.showMenu === true
              ?
              <IconButton
                className="appbar--menu-toggler"
                contrast
                onTouchTap={this.props.onMenuClick}
              >
                <MenuIcon />
              </IconButton>
              :
              null
            }

            <Typography type="title" colorInherit >MODXCLUB</Typography>

          </div>

          <Layout
            container
            gutter={0}
            justify="flex-end"
            align="center"
          >

            <Search />


            <Proxy store={this.props.store}/>
            <Notifications />
            <Balance user={this.props.user}/>
            <ProfileIcon />
          </Layout>
        </AppBar>
      </div>
    );
  }
}

 
export default MainAppBar
