// @modxclub/old__/fields/user/link.js
import React, {Component} from 'react';

import PropTypes from 'prop-types';
 
import {Link} from 'react-router-dom';
import Avatar from '../../Avatar';

export default class UserLink extends Component{

	static propTypes = {
		user: PropTypes.object.isRequired,
	}

	constructor(props){

		super(props);

		this.state = {}
	}


	render(){

		let {
			user,
			...other
		} = this.props;

		let {
			id,
			username,
			photo,
			author_avatar,
			fullname,
			blocked,
			active,
		} = user || {};

		if(!id){

			return null;
		}

		return <Link 
      to={`/profile/${username}`}
      href={`/profile/${username}`}
      className="underline-none"
      style={{
        display:"flex",
        flexDirection:"row",
        alignItems: "center",
      }}
      {...other}
    >
      <Avatar
        avatar={photo || author_avatar}
        username={fullname}
      />

      <div
        style={{
          whiteSpace: 'nowrap',
          marginLeft: 5,
        }}
      >
        {blocked == "1" || active != "1"
          ?
            <s>
              {fullname || username}
            </s>
          :
            fullname || username
        }
      </div>
    </Link>;
	}
}
