// @modxclub/old__/search/index.jsx
import React, {Component} from 'react';

import PropTypes from 'prop-types';

// import Typography from 'material-ui/Typography';

import {Link} from 'react-router-dom';

import SearchIcon from 'material-ui-icons/Search';
import IconButton from 'material-ui/IconButton';

import AutoCompletePrototype from 'material-ui-components/src/AutoComplete';


class AutoComplete extends AutoCompletePrototype{



  loadData(query){

    if(!query || query.length < 3){
      return false;
    }

    return super.loadData(query);
  }

  // componentDidMount(){
  // }



  render(){

    let {
      dataSource,
    } = this.state;

    if(dataSource && dataSource.length){
      dataSource.map(item => {
        if(!item.formattedName){
          item.formattedName = <span
            style={{
              color: '#333',
            }}
          >
            <div>
              {item.pagetitle}
            </div>

            <span
              style={{
                // fontStyle: 'italic',
                fontWeight: 'bold',
                fontSize: 14,
              }}
            >
              {item.blog_pagetitle}
            </span>
          </span>
        }
      });
    }

    // console.log('dataSource', dataSource);

    return super.render();
  }
}


export default class SearchField extends Component{

  static contextTypes = {
    connector_url: PropTypes.string.isRequired,
    documentActions: PropTypes.object.isRequired,
  }

  constructor(props){

    super(props);

    this.state = {
      open: false,
    }
  }

  componentWillMount(){

  }

  componentDidMount(){

  }

  componentDidUpdate(){

    // if(this.props.debug){
    //   console.log("MyComponent componentDidUpdate", this);
    // }
  }

  render(){

    let {
      documentActions,
      connector_url,
    } = this.context;

    let {
      open,
    } = this.state;

    return <Link
      to="/topics/"
      href="/topics/"
    >
      <IconButton>
        <SearchIcon
          color="#fff" 
        />
      </IconButton>
    </Link>
 
  }
}
