// @modxclub/old__/view/contacts/index.jsx
import React, { Component } from 'react';
import CallbackForm from '../../forms/callbackform'

import PropTypes from 'prop-types';

import Grid from 'material-ui/Grid';
import Paper from 'material-ui/Paper';
import Typography from 'material-ui/Typography';

import { createStyleSheet } from 'jss-theme-reactor';
import customPropTypes from 'material-ui/utils/customPropTypes';

import {browserHistory} from 'react-router-dom';

// window.browserHistory = browserHistory;

const defaultProps = {};

const styleSheet = createStyleSheet('ContactsView', (theme) => ({
  root: {

    '& .paper': {
      // height: '100%',
      marginBottom: 20,
    },
  },
  paper: {
    padding: 15,
  },
}));

var classes;

export default class ContactsView extends Component {

  static contextTypes = {
    styleManager: customPropTypes.muiRequired,
    setPagetitle: PropTypes.func.isRequired,
  }

  componentWillMount(){
    classes = this.context.styleManager.render(styleSheet);
  }

  componentDidMount(){

    let {
      setPagetitle,
    } = this.context;

    setPagetitle("Контакты");
  }

  render() {
    return (
    <div>
        <Typography type="title">Контакты</Typography>

        <Grid
          container
          style={{
            marginTop: 30,
          }}
          className={classes.root}
        >

          <Grid
            item
            xs={12}
            lg={6}
            xl={8}
          >
          
            {/*
            <Paper className={[classes.paper, "paper"].join(" ")}>
              <p>
                Уважаемые клиенты. 
              </p>
              <p>
                Если вы хотите заказать разработку сайта или сервиса, сообщаем, что до сентября мы заняты. 
              </p>
              <p>
                Если у вас не горящий проект и вы готовы внести предоплату сейчас, чтобы встать в график, можете присылать запрос.
              </p>
            </Paper>
            */}
            
            <Paper className={[classes.paper, "paper"].join(" ")}>
              <h4 className="header2"><a href="javascript:;">MODX-Клуб</a> - это высокопрофессиональная команда веб-разработчиков.</h4>
              <p>Мы разрабатываем проекты практически любого уровня, от простых сайтов-визиток до крупных интернет-магазинов, порталов и
                закрытых бизнес-решений.</p>
              <p>ООО "Технология"</p>
              <p>ОГРН: <strong>1157746742120</strong></p>
              <p>ИНН: <strong>7704326153</strong></p>
              <a href="https://elba.kontur.ru/card/f3kqtii8h6" target="_blank">Полные реквизиты</a>
              <p>Если вы хотите заказать интернет-магазин, сайт, или вам требуется поддержка вашего сайта - <a
                href="mailto:sales@modxclub.ru">sales@modxclub.ru</a></p>
              <p>Если вы заметили ошибку на сайте, или у вас возникли какие-то сложности в работе с ним - <a
                href="mailto:support@modxclub.ru">support@modxclub.ru</a></p>
              <p>Бухгалтерия: <a href="mailto:buh@modxclub.ru">buh@modxclub.ru</a></p>
              <p>По всем другим вопросам - <a href="mailto:office@modxclub.ru">office@modxclub.ru</a></p>
            </Paper>
          </Grid>

          <Grid
            item
            xs={12}
            lg={6}
            xl={4}
          >
            <CallbackForm />
          </Grid>

        </Grid>



    </div>


    );

    return null;
  }


}


ContactsView.defaultProps = defaultProps;
