// @modxclub/ui/src/Link/Tag/index.js
import React, { Component, Fragment } from 'react'
import PropTypes from 'prop-types'
import Typography from "material-ui/Typography";


import Link from '../';

import { withStyles } from 'material-ui/styles';

const styles = theme => {

  const {
    palette: {
      // background: {
      // },
      primary,
      grey,
      getContrastText,
    },
  } = theme;

  // const backgroundColor = primary[500];
  const backgroundColor = grey[200];
  const color = getContrastText(backgroundColor);

  const backgroundColorActive = primary[600];
  const colorActive = getContrastText(backgroundColorActive);

  // const backgroundColorHover = "#39a71f";
  const backgroundColorHover = primary[500];
  const colorHover = getContrastText(backgroundColorHover);

  return {
    root: {
    },
    text: {
      fontSize: "0.8rem",
      display: "inline-block",
      backgroundColor,
      color,
      borderRadius: 3,
      margin: 3,
      padding: "1px 3px",

      "&.active": {
        backgroundColor: backgroundColorActive,
        color: colorActive,
      },

      "&:hover": {
        backgroundColor: backgroundColorHover,
        color: colorHover,
      },
    },
  }
};


export class TagLink extends Component {

  render() {

    const {
      object,
      children,
      classes,
      ...other
    } = this.props;


    if (!object) {
      return null;
    }

    const {
      name,
    } = object;


    if (!name) {
      return null;
    }

    return <Link
      to={`/tag/${name}`}
      title={name}
      classes={classes}
      {...other}
    >
      {name}
    </Link>
  }
}


export default withStyles(styles)(TagLink);
