// @modxclub/old__/articles/articleinfocommentsform.jsx
import React from 'react';
 

import Button from 'material-ui/Button';
import TextField from 'material-ui/TextField';

import IconSend from 'material-ui-icons/Send';

class ArticleInfoCommentsForm extends React.Component {

  // create comment
  handleAddComment(){
    // var comments = this.props.comments;
    var comment = {
      id: new Date().getTime(),
      createdon: '20 февраля 2017 в 04:52',
      author_avatar: '',
      author_username: 'Snow',
      comment_text: this.refs.commentField.getValue(),
    };
    //
    // object.comments.push(comment);
    // this.props.ArticleInfo.setState({
    //   comments: object.comments
    // });
    //
    this.refs.commentField.getInputNode().value = '';

    this.props.addComment(comment);
  }

  //{/* value={comment ? <a href={'#comment-'+ comment.idea}>{comment.author_username}</a> : ''} */}
  render(){
    let comment = this.props.comment;
// console.log('ArticleInfoCommentsForm',comment);
    return (
      <div>
        <div>
          <TextField
            ref="commentField"
            hintText="Напишите ваш комментарий"
            multiLine={true}
            fullWidth={true}
            label="Комментарий" />
        </div>
        <Button
          icon={<IconSend />}
          onTouchTap={this.handleAddComment.bind(this)}
        >Комментировать</Button>
      </div>
    );
  }
}
 

export default ArticleInfoCommentsForm
