// @modxclub/old__/fields/blogs/index.js
import React, {Component} from 'react';

import PropTypes from 'prop-types';

import Grid from 'material-ui/Grid';
import Checkbox, {LabelCheckbox} from "material-ui/Checkbox";

// import AutoComplete from 'material-ui-components/src/AutoComplete';

const defaultProps = {}

export default class BlogsSelectField extends Component{

	// static contextTypes = {
	// 	request: PropTypes.func.isRequired,
	// };

	static propTypes = {
		blogs: PropTypes.array.isRequired,
	}

	constructor(props){

		super(props);

		this.state = {
			// dataSource: [],
		}
	}

	componentWillMount(){

	}

	// componentDidMount(){

	// 	this.loadData();
	// }

  // componentDidUpdate(){

  //   if(this.props.debug){
  //     console.log("MyComponent componentDidUpdate", this);
  //   }
  // }

  onChange(event){

  	let {
  		onChange,
  	} = this.props;

  	let {
  		name,
  		value,
  		checked,
  	} = event.target;

  	// console.log('onChange', name, value, checked);

  	onChange && onChange(event);
  }

	render(){

		let {
		} = this.state;

		let{
			blogs: dataSource,
		} = this.props;

		let {
			checkedBlogs,
		} = this.props;

		// console.log("checkedBlogs", checkedBlogs);

		if(!dataSource || !dataSource.length){
			return null;
		}

		let blogs = [];

		dataSource.map(blog => {

			let {
				id,
				pagetitle,
			} = blog;

			// let name = `blog[${id}]`;
			let name = id;

			blogs.push(<Grid
				key={id}
				item
				xs={12}
				sm={6}
				md={4}
				lg={3}
			>
				<LabelCheckbox
	        label={blog.pagetitle}
	        checked={checkedBlogs && checkedBlogs.find(n => n == id) ? true : false}
	        name={name}
	        value={pagetitle}
	        onChange={event => {
	          this.onChange(event);
	        }}
	      />
			</Grid>);
		});

		return <Grid
			container
		>
			{blogs}
		</Grid>

		// return <AutoComplete
		// 	label="Раздел"
		// 	dataSource={dataSource}
		// />;
	}
}
