// @modxclub/old__/balance/index.jsx
import React, {Component} from 'react';

import BalanceIcon from 'material-ui-icons/AccountBalanceWallet';
import Chip from 'material-ui/Chip';
import {Link} from 'react-router-dom';

const defaultProps = {
};

export default class Balance extends Component {

  state = {
  };


  constructor(props) {

    super(props);

  }





  componentWillReceiveProps(nextProps){

    return true;
  }


  // componentWillMount(){
  //   this.state.sending = true;
  // }

  ShowBalanceDetails(){

  }

  render() {

    var icon;

    let {
      user,
    } = this.props;

    let {
      username,
    } = user || {};

    if(!user || user.id == 0 || !username){
      return null;
    }

    var balance = this.props.user.balance || 0;

    icon =
        <Link
          to={`/profile/${username}/`}
          href={`/profile/${username}/`}
          style={{
            color:'#fff',
            textDecoration: 'none',
          }}
        >
          <BalanceIcon
            style={{
              verticalAlign: 'middle',
            }}
          />
          {balance}
        </Link>;


    return icon;
  }
}


Balance.defaultProps = defaultProps;
