// @modxclub/old__/view/office/sections/layout.js
import React, {Component} from 'react';

import PropTypes from 'prop-types';

import IndexView from '../../index/';

import Grid from 'material-ui/Grid';
import Typography from 'material-ui/Typography';

import List, {
  ListItem,
  ListItemIcon,
  ListItemText,
  ListItemSecondaryAction,
  ListSubheader,
} from 'material-ui/List';

import {Link} from 'react-router-dom';

// console.log('IndexView', IndexView);

export default class OfficeSectionLayout extends IndexView{
 

	constructor(props){

		super(props);

		Object.assign(this.state, {
      
    });
	}

	getMenu(){

    let {
      user: {
        user,
      }
    } = this.props;

    let {
      username,
    } = user || {};

    return <List> 

      <Link
        to={`/office/`}
        href={`/office/`}
      >
        <ListItem
          button
        > 
          В начало
        </ListItem>
      </Link>
    </List>
  }



  getPageHeader(pagetitle){

    return super.getPageHeader(pagetitle || "Управление командами");
  }

  setPagetitle(title){
 

    return super.setPagetitle(title || "Управление командами");
  }

  renderContent(){

  }

 	contentBlock(){


    return <div>
    	

    	<Grid
	      container
	    >
	      <Grid
	        item
	      >
    			{this.getMenu()}
	      </Grid>

	      <Grid
	        item
	        xs={12}
	      >
	        {this.renderContent()}
	      </Grid>
	      
	    </Grid>
    </div>;
 	}
}
