// @modxclub/old__/redux/actions/documentActions.js
export const INITIALIZATION_REQUESTED = 'INITIALIZATION_REQUESTED';
export const INFORMER_MESSAGE_ADDED = 'INFORMER_MESSAGE_ADDED';
export const INFORMER_MESSAGE_REMOVED = 'INFORMER_MESSAGE_REMOVED';
export const INFORMER_MESSAGE_SHOWED = 'INFORMER_MESSAGE_SHOWED';
export const LOAD_DOCUMENT_REQUESTED = 'LOAD_DOCUMENT_REQUESTED';
export const DOCUMENT_LOADED = 'DOCUMENT_LOADED';
export const VIEW_SWITCHED = 'VIEW_SWITCHED';

export function InitializeDocument() {
  return {
    type: INITIALIZATION_REQUESTED,
  };
}

export function addInformerMessage(message) {

  return {
    type: INFORMER_MESSAGE_ADDED,
    message,
  };
}

export function removeInformerMessage(message) {

  return {
    type: INFORMER_MESSAGE_REMOVED,
    message,
  };
}

export function InformerMessageShowed() {
  console.error("Method InformerMessageShowed depricated and will be removed in future");
  return {
    type: INFORMER_MESSAGE_SHOWED,
  };
}

export function LoadDocument(url) {

  return {
    type: LOAD_DOCUMENT_REQUESTED,
    url,
  };
}

export function DocumentLoaded(document) {

  return {
    type: DOCUMENT_LOADED,
    document,
  };
}

/*
 * */
export function SwitchView(view, data) {
  return {
    type: VIEW_SWITCHED,
    view,
    data,
  };
}
