// @modxclub/old__/AutoComplete/index.jsx
import React, {Component} from 'react';

import { createStyleSheet } from 'jss-theme-reactor';
import customPropTypes from 'material-ui/utils/customPropTypes';

import TextField from 'material-ui/TextField';

// import Popover from 'material-ui/internal/Popover';
import Popover from '../internal/Popover';

import _ from 'lodash';

import List, {
  ListItem,
  ListItemText,
  ListItemIcon,
  ListItemSecondaryAction
} from 'material-ui/List';

const defaultProps = {
  input_id: _.uniqueId('autocomplete_'),
  popover_id: _.uniqueId('popover_'),
  label: "",
  maxSearchResults: 10,
  searchText:"",
  dataSource:[],
  data:[],
  anchorEl: null,
}

const styleSheet = createStyleSheet('AutoComplete', () => ({
  menu: {},
  content: {
  },
  listItem: {
    paddingTop: 6,
    paddingBottom: 6,
  },
}));

var classes;

export default class AutoComplete extends Component{

  constructor(props){

    super(props);


    this.state = {
      input_id: props.input_id,
      popover_id: props.popover_id,
      searchText: props.searchText,
      open: false,
      // dataSource: props.dataSource,
    };
  }

  componentWillMount(){

    classes = this.context.styleManager.render(styleSheet);
  }

  onChange(event){

    var value = event.target.value;

    var result = [];

    if(
      this.props.dataSource
      && value != ""
    ){
      for(var i in this.props.dataSource){

        var item = this.props.dataSource[i];

        if(!item){
          continue;
        }

        if(item.match(value)){
          result.push(item);

          if(this.props.maxSearchResults > 0 && this.props.maxSearchResults == result.length){

            break;
          }
        }
      }
    }

    this.setState({
      searchText: value,
      data: result,
      anchorEl: event.target,
      open: true,
    });
  }


  render(){

    var items = [];

    if(this.state.data && this.state.data.length){
      this.state.data.map((item) => {
        items.push(<ListItem
          key={item}
          className={classes.listItem}
          button
          onTouchTap={() => {

            if(this.props.onNewRequest){
              this.props.onNewRequest(item);
            }

            this.setState({
              searchText: "",
              open: false,
            });
          }}
        >
          <ListItemText primary={item} />
        </ListItem>);
      });
    }

    return <div>
      <TextField
        aria-owns={this.state.popover_id}
        aria-haspopup="true"
        label={this.props.label}
        error={this.props.error}
        value={this.state.searchText}
        onChange={(event) => {
          this.onChange(event);
        }}
        style={{
          zIndex: 1000,
        }}
      />

      <Popover
        id={this.state.popover_id}
        anchorEl={this.state.anchorEl}
        open={this.state.open && items && items.length > 0}
        onRequestClose={this.handleRequestClose}
        modal={false}
        anchorOrigin={{
          vertical: 'bottom',
          horizontal: 'left',
        }}
      >
        <List>
          {items}
        </List>

      </Popover>
    </div>
  }
}

AutoComplete.defaultProps = defaultProps;

AutoComplete.contextTypes = {
  styleManager: customPropTypes.muiRequired,
};
