// @modxclub/old__/view/companies/company/index.js
import React from 'react';

import IndexView from '../../index/';

import Grid from 'material-ui/Grid';
import Button from 'material-ui/Button';
import Typography from 'material-ui/Typography';

import {Link} from 'react-router-dom';

import WorksList from '../../works/list';
import ProfilesList from '../../profiles/list';

export default class CompanyView extends IndexView {

  constructor(props){

    super(props);


    let {
      data,
    } = props.document.document || {};

    let company = data && data.object || {};

    Object.assign(this.state, {
      company,
    });

  }

  componentDidMount(){

  }

  getPageHeader(pagetitle){

    let {
      company
    } = this.state;

    pagetitle = pagetitle || `Компания "${company.pagetitle}"`;

    return super.getPageHeader(pagetitle);
  }

  contentBlock(){

    // let {
    //   company,
    // } = this.state;


    let {
      data,
    } = this.props.document.document || {};

    let company = data && data.object || {};

    let {
      id: company_id,
      pagetitle,
      works: works_response,
      members: members_response,
      company_resource_id,

    } = company;

    let {
      object: works,
      total: works_total,
    } = works_response || {}

    // console.log('document', pagetitle, company);
    // console.log('company members', members_response, company);

    let {
      object: profiles,
      total: profiles_total,
      page: profiles_page,
      limit: profiles_limit,
    } = members_response || {}

    return <Grid
      container
    >
      
      <Grid
        item
        xs={12}
      >

        <Typography>
          Контактная информация
        </Typography>
        
      </Grid>
      
      {profiles && Array.isArray(profiles) && profiles.length
        ?
          <Grid
            item
            xs={12}
          >
            <Typography 
              type="title"
              style={{
                // marginLeft: 17,
                marginBottom: 5,
              }}
            >
              Команда
            </Typography>

            <ProfilesList 
              profiles={profiles}
              page={profiles_page}
              limit={profiles_limit}
              total={profiles_total}
            />
            
          </Grid>
        :
        null}

      
      <Grid
        item
        xs={12}
      >

        <Typography> 
        </Typography>
        
      </Grid>
      
      <Grid
        item
      >

        <Typography>
          
        </Typography>
        
      </Grid>
      
      <Grid
        item
      >

        <Typography>
          
        </Typography>
        
      </Grid>
      
      <Grid
        item
      >

        <Typography>
          
        </Typography>
        
      </Grid>
      
      <Grid
        item
      >

        <Typography>
          
        </Typography>
        
      </Grid>

      {works && works.length
        ?
          <Grid
            item
            xs={12}
          >

            <Typography 
              type="title"
              style={{
                // marginLeft: 17,
                marginBottom: 5,
              }}
            >
              Портфолио
            </Typography>
            
            <WorksList
              works={works}
              page={0}
              limit={0}
              total={works.length}
            />


            {/*
              Если работ более 4, отправляем в полное портфолио
            */}

            {works_total > 4 
              ?
                <div>

                  <Link
                    to={`katalog-sajtov/?company=${company_id}`}
                    href={`katalog-sajtov/?company=${company_id}`}
                    className="underline-none"
                  >
                    <Button>
                      Смотреть все работы команды
                    </Button>
                  </Link>

                  <Link
                    to={`katalog-sajtov/?developer=${company_id}`}
                    href={`katalog-sajtov/?developer=${company_id}`}
                    className="underline-none"
                  >
                    Только собственные проекты
                  </Link>

                </div>
              :
              null
            }

          </Grid>
        :
        null
      }

    </Grid>
  }
}
