// @modxclub/old__/view/index/index.jsx
import React, {Component} from 'react';

import PropTypes from 'prop-types';

// import ReactDOM from 'react-dom';

import Typography from 'material-ui/Typography';


// import Articles from '../../articles/index';


const defaultProps = {
};



export default class IndexView extends Component {

  static contextTypes = {
    setPagetitle: PropTypes.func.isRequired,
  }

  constructor(props){

    super(props);

    this.state = {};
  }

  componentDidMount(){

    this.setPagetitle();

    return;
  }

  componentDidUpdate(prevProps, prevState){

    this.setPagetitle();

    return;
  }

  showUserProfile(username){
    this.props.documentActions.LoadDocument("profile/" + username);
  }

  getPageHeader(pagetitle){

    let {
      pageHeader,
    } = this.state;

    pagetitle = pagetitle || pageHeader;

    return pagetitle && <Typography
      type="headline"
      className="pageHeader"
    >{pagetitle}</Typography> || null;
  }

  setPagetitle(title){


    let {
      pagetitle,
    } = this.state;

    let {
      setPagetitle,
    } = this.context;

    title = title || pagetitle;

    if(!title){
      return;
    }

    setPagetitle && setPagetitle(title);
  }

  contentBlock(){}

  render(){
    var document = this.props.document || {};

    return (<div>

      {this.getPageHeader()}

      <div dangerouslySetInnerHTML={{ __html: document.document.content}} />

      {this.contentBlock()}

    </div>);
  }
}


IndexView.defaultProps = defaultProps;
